google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,331 @@
+
+  /**
+   * The last document snapshot returned by the stream. Used to set the startAfter() field in the
+   * subsequent stream.
+   */
+  @Nullable private DocumentSnapshot lastDocumentSnap;
+
+  /**
+   * The number of pending BulkWriter operations. Used to determine when the next query can be run.
+   */
+  private int pendingOperationsCount = 0;
+
+  RecursiveDelete(
+      FirestoreRpcContext<?> firestoreRpcContext,
+      BulkWriter writer,
+      ResourcePath path,
+      int maxLimit,
+      int minLimit) {
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.path = path;
+    this.maxPendingOps = maxLimit;
+    this.minPendingOps = minLimit;
+  }
+
+  public ApiFuture<Void> run() {
+    Preconditions.checkState(!started, "RecursiveDelete.run() should only be called once");
+    started = true;
+
+    writer.verifyNotClosed();
+
+    streamDescendants();
+    return completionFuture;
+  }
+
+  private void streamDescendants() {
+    Query query = getAllDescendantsQuery();
+    synchronized (lock) {
+      streamInProgress = true;
+    }
+    final int[] streamedDocsCount = {0};
+    final ApiStreamObserver<DocumentSnapshot> responseObserver =
+        new ApiStreamObserver<DocumentSnapshot>() {
+          public void onNext(DocumentSnapshot snapshot) {
+            streamedDocsCount[0]++;
+            lastDocumentSnap = snapshot;
+            deleteReference(snapshot.getReference());
+          }
+
+          public void onError(Throwable throwable) {
+            String message = "Failed to fetch children documents";
+            synchronized (lock) {
+              lastError =
+                  FirestoreException.forServerRejection(Status.UNAVAILABLE, throwable, message);
+            }
+            onQueryEnd();
+          }
+
+          public void onCompleted() {
+            synchronized (lock) {
+              streamInProgress = false;
+            }
+            // If there are fewer than the number of documents specified in the limit() field, we
+            // know that the query is complete.
+            if (streamedDocsCount[0] < maxPendingOps) {
+              onQueryEnd();
+            } else if (pendingOperationsCount == 0) {
+              // Start a new stream if all documents from this stream were deleted before the
+              // `onCompleted()` handler was called.
+              streamDescendants();
+            }
+          }
+        };
+
+    query.stream(responseObserver);
+  }
+
+  private Query getAllDescendantsQuery() {
+    ResourcePath parentPath;
+    String collectionId;
+    if (path.isDocument()) {
+      // The parent is the closest ancestor document to the location we're deleting. Since we are
+      // deleting a document, the parent is the path of that document.
+      parentPath = path;
+      collectionId = path.getParent().getId();
+      Preconditions.checkState(collectionId != null, "Parent of a document should not be null.");
```

Comment:
I would drop this check - every document has a collection ID. If you do keep it, the check should be on getParent(), which is nullable and not on getId(). 

BTW, I really like this cleanup.

Reply:
Changed check to parent.

I definitely like this version more, thanks for pushing through on your vision :) Having all the collectionId/parentPath logic inside getAllDescendantsQuery() makes so more sense, and makes the rest of the code much cleaner!